File: 46. Permutations/Solution.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def backTrack(path, choices):
            if not choices:
                result.append(path[:])
                
            for i in range(len(choices)):
                choice = choices.pop(i)
                path.append(choice)
                backTrack(path, choices)
                path.pop()
                choices.insert(i, choice)
        
        result = []
        backTrack([], nums)
        return result

File: 46. Permutations/Solution.java
class Solution {
    List<List<Integer>> res = new LinkedList();
    
    public List<List<Integer>> permute(int[] nums) {
        LinkedList<Integer> track = new LinkedList();
        backTrack(nums, track);
        return res;
    }
    
    void backTrack(int[] nums, LinkedList<Integer> track) {
        if (track.size() == nums.length) {
//             what if I don't use new
            res.add(new LinkedList(track));
            return;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (track.contains(nums[i])) continue;
            track.add(nums[i]);
            backTrack(nums, track);
            track.removeLast();

        }
    }
}